### ISSUE TITLE
Enforcing minimum number of instances doesn't work

### ISSUE THREAD

**Issue by Panaetius** [RDFLib/pySHACL#36]
Following the trick mentioned in https://www.w3.org/wiki/SHACL/Examples i wanted to write a shape to validate the existence of a node.

The shape 
```
{
    "@context": {
       "rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#",
       "sh": "http://www.w3.org/ns/shacl#",
       "schema": "http://schema.org/"
    },
    "@graph": [
        {
            "@id": "_:forceDatasetShape",
            "@type": "sh:NodeShape",
            "sh:targetNode": "schema:DigitalDocument",
            "sh:property": [
                {
                    "sh:path": [
                        {
                            "sh:inversePath": [{
                                "@id": "rdf:type",
                                "@type": "@id"
                             }]
                        }
                    ],
                    "sh:minCount": 1
                }
            ]
        }
    ]
}
```

with the graph
```
{}
```

throws a validation error in the SHACL playground https://shacl.org/playground/

But pyshacl says that it's conforming. Does this `inversePath` trick not work with pySHACL?

Command I'm using is: `pyshacl -a -m -s shape.json graph.json -sf json-ld -df json-ld`
With pyshacl version 0.11.3

--------------------------
On a side note, SHACL playground validates successfully with
```
{
    "@context": { "schema": "http://schema.org/", "rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#" },

    "@id": "http://example.org/ns#Bob",
    "rdf:type": "http://schema.org/DigitalDocument"
}
```

but not with 
```
{
    "@context": { "schema": "http://schema.org/", "rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#" },

    "@id": "http://example.org/ns#Bob",
    "@type": "http://schema.org/DigitalDocument"
}
```

 or

```
{
    "@context": { "schema": "http://schema.org/", "rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#" },

    "@id": "http://example.org/ns#Bob",
    "rdf:type": "schema:DigitalDocument"
}
```

which is weird. I was under the impression that `@type` is an alias for `rdf:type`.

**Comment by ashleysommer** [comment]
Hi @Panaetius 

> with the graph
>```
>{}
>```

Did you mean to show an empty Data Graph here?
In PySHACL, an empty graph will always validate "True", as there's nothing which is non-conforming.

>Does inversePath not work with pySHACL?

PySHACL supports all SHACL property paths, including `sh:inversePath`. See the list of supported features here: https://github.com/RDFLib/pySHACL/blob/master/FEATURES.md

However, to get to the crux of the issue; you have pointed out one design flaw in the way PySHACL applies shape's constraints to target nodes. PySHACL will always first try to determine which nodes in the Graph this shape will apply to. It does this by checking for `targetClass`, `targetNodes`, and "implicit classes" that a Shape can work with. It then tries to find nodes which match these rules in the Data Graph. Finally, it applies the shape's constraints to those nodes.

What is happening here, is the Shape's `targetNode` is specifying `schema:DigitalDocument` as the node to focus on, but there is no `schema:DigitalDocument` in the Data Graph, so pySHACL simply skips this Shape because it has no Focus node.

At the step where pySHACL searches for Focus nodes for a Shape to apply to, there is no way for PySHACL to know the property path is going to be inverted.

I have an idea for a possible refactor of this part of the code, which will make it possible for this snippet to work as expected. But that will probably be released as part of the v0.12.x cycle.

**Comment by Panaetius** [comment]
Thank you for the quick reply! 

Yes, I meant to show an empty graph. 

I guess a better way to describe the empty graph would be 
```
{
    "@context": { "@vocab": "http://www.schema.org/" },
    "@graph": {}
}
```

So `schema:DigitalDocument` is there in the vocabulary, but not used in the graph itself. I understand this would still conform at the moment.

Thanks for looking into it!

Do you know of any workaround in SHACL that would allow forcing that a node of a specific type has to exist somewhere in the graph without specifying where?

**Comment by ashleysommer** [comment]
> Do you know of any workaround in SHACL that would allow forcing that a node of a specific type has to exist somewhere in the graph without specifying where?

No, actually I don't know a way of doing this.

I was reading over the SHACL Spec again, and in section [2.1.3 Targets](https://www.w3.org/TR/shacl/#targets) I found this line:
> RDF terms produced by targets are not required to exist as nodes in the data graph

So there's a section of the code in PySHACL which violates that part of the spec. What surprises me is there are no tests in the SHT nor the DASH test suites to ensure this behaviour, because PySHACL passes both of those SHACL test suites.

I am going to add a couple of new tests to PySHACL to ensure this functionality, they will fail for now but hopefully soon in the future the behaviour will be fixed and they will start passing.

**Comment by ashleysommer** [comment]
Hi @Panaetius

Ok, good news. That was easier than I thought. I simply removed a section of code which enforces that TargetNodes must exist in the DataGraph. And the tests started passing. And no regressions.

I've released a new bugfix version (in the v0.11.3 series) which contains this fix, its called v0.11.3.post1 it should be available on pypi in about 5 minutes.

**Comment by Panaetius** [comment]
Thank you very much :tada: 

Works like a charm.
